tools/clang/lib/SPIRV/EmitVisitor.cpp

```diff
@@ -17,14 +17,59 @@
 namespace clang {
 namespace spirv {
 
+void EmitVisitor::emitDebugNameForInstruction(uint32_t resultId,
+                                              llvm::StringRef debugName) {
+  // Most instructions do not have a debug name associated with them.
+  if (debugName.empty())
+    return;
+
+  curInst.clear();
+  curInst.push_back(static_cast<uint32_t>(spv::Op::OpName));
+  curInst.push_back(resultId);
+  encodeString(debugName);
+  curInst[0] |= static_cast<uint32_t>(curInst.size()) << 16;
+  debugBinary.insert(debugBinary.end(), curInst.begin(), curInst.end());
+}
+
 void EmitVisitor::initInstruction(spv::Op op) {
   curInst.clear();
   curInst.push_back(static_cast<uint32_t>(op));
 }
 
 void EmitVisitor::finalizeInstruction() {
+  const auto op = static_cast<spv::Op>(curInst[0]);
   curInst[0] |= static_cast<uint32_t>(curInst.size()) << 16;
-  spirvBinary.insert(spirvBinary.end(), curInst.begin(), curInst.end());
+  switch (op) {
+  case spv::Op::OpCapability:
+  case spv::Op::OpExtension:
+  case spv::Op::OpExtInstImport:
+  case spv::Op::OpMemoryModel:
+  case spv::Op::OpEntryPoint:
+  case spv::Op::OpExecutionMode:
+  case spv::Op::OpExecutionModeId:
+    preambleBinary.insert(preambleBinary.end(), curInst.begin(), curInst.end());
+  case spv::Op::OpString:
+  case spv::Op::OpSource:
+  case spv::Op::OpSourceExtension:
+  case spv::Op::OpSourceContinued:
+  case spv::Op::OpName:
+  case spv::Op::OpMemberName:
+  case spv::Op::OpModuleProcessed:
+    debugBinary.insert(debugBinary.end(), curInst.begin(), curInst.end());
+  case spv::Op::OpDecorate:
+  case spv::Op::OpDecorateId:
+  case spv::Op::OpMemberDecorate:
+  case spv::Op::OpGroupDecorate:
+  case spv::Op::OpGroupMemberDecorate:
+  case spv::Op::OpDecorationGroup:
+  case spv::Op::OpDecorateStringGOOGLE:
+  case spv::Op::OpMemberDecorateStringGOOGLE:
+    annotationsBinary.insert(annotationsBinary.end(), curInst.begin(), curInst.end());
+  case spv::Op::OpLine:
```

Comment:
OpLine can be everywhere. They should just go to the `default` case.

Reply:
Done